File: island_sim/src/simulator/world.py
"""WorldState utility methods for querying and modifying the grid."""

from __future__ import annotations

import numpy as np

from ..constants import (
    BUILDABLE_TERRAINS,
    FOREST,
    MOUNTAIN,
    OCEAN,
    PLAINS,
    PORT,
    RUIN,
    SETTLEMENT,
)
from ..types import Settlement, WorldState


def get_neighbors(
    state: WorldState, x: int, y: int, radius: int = 1
) -> list[tuple[int, int]]:
    """Return all grid coordinates within Chebyshev distance `radius` of (x, y), excluding (x, y) itself."""
    result = []
    for dy in range(-radius, radius + 1):
        for dx in range(-radius, radius + 1):
            if dx == 0 and dy == 0:
                continue
            nx, ny = x + dx, y + dy
            if 0 <= nx < state.width and 0 <= ny < state.height:
                result.append((nx, ny))
    return result


def get_adjacent_terrain_counts(state: WorldState, x: int, y: int) -> dict[int, int]:
    """Count terrain types in the 8 cells adjacent to (x, y)."""
    counts: dict[int, int] = {}
    for nx, ny in get_neighbors(state, x, y, radius=1):
        terrain = int(state.grid[ny, nx])
        counts[terrain] = counts.get(terrain, 0) + 1
    return counts


def is_coastal(state: WorldState, x: int, y: int) -> bool:
    """Check if position (x, y) is adjacent to ocean."""
    for nx, ny in get_neighbors(state, x, y, radius=1):
        if state.grid[ny, nx] == OCEAN:
            return True
    return False


def distance(x1: int, y1: int, x2: int, y2: int) -> float:
    """Chebyshev distance between two points."""
    return max(abs(x1 - x2), abs(y1 - y2))


def find_settlements_in_range(
    state: WorldState, x: int, y: int, radius: float, alive_only: bool = True
) -> list[Settlement]:
    """Find all settlements within Chebyshev distance of (x, y)."""
    result = []
    for settlement in state.settlements:
        if alive_only and not settlement.alive:
            continue
        if distance(x, y, settlement.x, settlement.y) <= radius:
            result.append(settlement)
    return result


def get_settlement_at(state: WorldState, x: int, y: int) -> Settlement | None:
    """Get the alive settlement at (x, y), if any."""
    for settlement in state.settlements:
        if settlement.alive and settlement.x == x and settlement.y == y:
            return settlement
    return None


def find_buildable_cells(
    state: WorldState, x: int, y: int, radius: int
) -> list[tuple[int, int]]:
    """Find all cells within radius that can receive a new settlement."""
    result = []
    for nx, ny in get_neighbors(state, x, y, radius):
        terrain = int(state.grid[ny, nx])
        if terrain in BUILDABLE_TERRAINS and get_settlement_at(state, nx, ny) is None:
            result.append((nx, ny))
    return result


def place_settlement(
    state: WorldState,
    x: int,
    y: int,
    owner_id: int,
    population: float = 1.0,
    food: float = 0.5,
    wealth: float = 0.0,
    defense: float = 0.3,
    tech_level: float = 0.0,
    has_port: bool = False,
) -> Settlement:
    """Place a new settlement at (x, y) and update the grid."""
    settlement = Settlement(
        x=x,
        y=y,
        population=population,
        food=food,
        wealth=wealth,
        defense=defense,
        tech_level=tech_level,
        has_port=has_port,
        has_longship=False,
        owner_id=owner_id,
        alive=True,
    )
    state.settlements.append(settlement)
    state.grid[y, x] = PORT if has_port else SETTLEMENT
    return settlement


def kill_settlement(state: WorldState, settlement: Settlement) -> None:
    """Mark a settlement as dead and convert its grid cell to a ruin."""
    settlement.alive = False
    state.grid[settlement.y, settlement.x] = RUIN

File: island_sim/rust/src/world.rs
/// Grid utility functions.

use crate::types::*;
use std::collections::HashMap;

/// Chebyshev distance between two points.
pub fn distance(x1: i32, y1: i32, x2: i32, y2: i32) -> f64 {
    (x1 - x2).abs().max((y1 - y2).abs()) as f64
}

/// Check if position is adjacent to ocean.
pub fn is_coastal(state: &WorldState, x: i32, y: i32) -> bool {
    for dy in -1..=1 {
        for dx in -1..=1 {
            if dx == 0 && dy == 0 { continue; }
            let nx = x + dx;
            let ny = y + dy;
            if nx >= 0 && nx < state.width && ny >= 0 && ny < state.height {
                if state.grid[ny as usize][nx as usize] == OCEAN {
                    return true;
                }
            }
        }
    }
    false
}

/// Count terrain types in the 8 cells adjacent to (x, y).
pub fn get_adjacent_terrain_counts(state: &WorldState, x: i32, y: i32) -> HashMap<i32, i32> {
    let mut counts = HashMap::new();
    for dy in -1..=1 {
        for dx in -1..=1 {
            if dx == 0 && dy == 0 { continue; }
            let nx = x + dx;
            let ny = y + dy;
            if nx >= 0 && nx < state.width && ny >= 0 && ny < state.height {
                let terrain = state.grid[ny as usize][nx as usize];
                *counts.entry(terrain).or_insert(0) += 1;
            }
        }
    }
    counts
}

/// Find all buildable cells within Chebyshev radius that don't have a settlement.
pub fn find_buildable_cells(state: &WorldState, x: i32, y: i32, radius: i32) -> Vec<(i32, i32)> {
    let mut result = Vec::new();
    for dy in -radius..=radius {
        for dx in -radius..=radius {
            if dx == 0 && dy == 0 { continue; }
            let nx = x + dx;
            let ny = y + dy;
            if nx < 0 || nx >= state.width || ny < 0 || ny >= state.height { continue; }
            let terrain = state.grid[ny as usize][nx as usize];
            if !is_buildable(terrain) { continue; }
            // Check no alive settlement at this position
            let occupied = state.settlements.iter().any(|s| s.alive && s.x == nx && s.y == ny);
            if !occupied {
                result.push((nx, ny));
            }
        }
    }
    result
}

/// Place a new settlement at (x, y) and update the grid.
pub fn place_settlement(
    state: &mut WorldState,
    x: i32, y: i32,
    owner_id: i32,
    population: f64,
    food: f64,
    wealth: f64,
    defense: f64,
    tech_level: f64,
    has_port: bool,
) {
    let settlement = Settlement {
        x, y, population, food, wealth, defense, tech_level,
        has_port, has_longship: false, owner_id, alive: true,
    };
    state.settlements.push(settlement);
    state.grid[y as usize][x as usize] = if has_port { PORT } else { SETTLEMENT };
}

/// Kill a settlement and convert its cell to ruin.
pub fn kill_settlement(state: &mut WorldState, idx: usize) {
    let s = &mut state.settlements[idx];
    s.alive = false;
    state.grid[s.y as usize][s.x as usize] = RUIN;
}
